### PR TITLE
Implement `--user` flag and user scheme support for `uv pip`

### LINKED ISSUE #2077: Support `--user` flag from pip
Currently (even in latest uv 0.1.12 that supports --python flag) it's not possible to use `uv` in case `--user` flag of `pip` (or PIP_USER="true" env variable is set). 

With `--user` flag, packages are installed to a ~/.local folder, which means that the user might not have access to the `python` system installation as a whole, but can locally install and uninstall packages, without having venv.

This is extremely useful for cases like `CI` and building container images, where having a venv is an extra overhead and it is unnecessary burden.

There is an interesting (and used in Apache Airflow) case for the `--user` flag (this currently prevent us from using `uv` by us and our users  also in PROD images in addition to our CI image). That is currently a blocker for https://github.com/apache/airflow/issues/37785

Why `--user` flag is useful and why we use it in Airflow? 

Using `--user` flag is pretty useful when you want to have an optimized image - because such `.local` folders can be copied between different stage of the image (based on the same base image and prod libraries installed) and you can copy the whole `.local` folder between the stages after packages are installed - which means that the final stage of the image does not have to have `build-essentials`/ compilers installed. 

You could do the same with `venv` if you keep it in the same folder, but that looses an essential capability of creating venv dynamically containing all the packages you have in your `.local` folder.  When you have your `--user` installed packages, and you create a new venv  with `--system-site-packages`,  the packages installed in `.local` folder are also installed in the new venv. This does not work when you create a new `venv` from another `venv` because `--system-site-packages` are only the ones installed in the system.

While I can think of some creative ways (maybe I will find some) - having an equivalent of `--user` installation by `uv pip install` would be a great simplification for our case to support the users who want to use `uv` (and seems that there is already a need for that looking at the https://github.com/apache/airflow/issues/37785

### PR DESCRIPTION
**Pull request by imfing**
## Summary

Hi team, excited for your awesome work of `uv`. 

This PR attempts to implement `--user` flag for `uv pip` subcommands: `install`, `freeze`, `list` and `show`.

Closes #2077 

also #1584

Per [PEP 370 – Per user site-packages directory](https://peps.python.org/pep-0370/), `pip` supports [user scheme package installation](https://pip.pypa.io/en/stable/user_guide/#user-installs) which would be useful for multi-user environment or CI use cases without the need of virtual environment or breaking system Python packages.

Example usage:

```shell
export PYTHONUSERBASE=~/.local/myappenv
uv pip install --user SomePackage
# SomePackage is available under ~/.local/myappenv
```

#### Implementation details

The implementation aims to bring minimal changes to the existing functionalities of `uv pip`, as well as supporting `--user` for pip compatibility.

Here are some breakdowns of the implementation:

* add [`_infer_user()`](https://github.com/pypa/pip/blob/ae5fff36b0aad6e5e0037884927eaa29163c0611/src/pip/_internal/locations/_sysconfig.py#L86) to `crates/uv-interpreter/src/get_interpreter_info.py` which returns the user scheme if specified
* add `PythonEnvironment::from_user_scheme()` which queries the interpreter info with user scheme, thus the `platlib`, `purelib`, `scripts`, `data` and `include` should point to the user site
* use the user scheme Python environment for `uv pip` subcommands when the `--user` flag is present, making the subcommands use *only* the user site for package search and installation.

Right now the implementation doesn't include the global packages mentioned in [User Installs](https://pip.pypa.io/en/stable/user_guide/#user-installs) since we only consider site packages from the `scheme.purelib`. I'm thinking about adding that later for better compatibility with `pip install`.

I am new to `uv`, looking forward to the team's feedback and suggestions :)

## Test Plan

May need some advice and guidance on what's the best practice on testing this.

### CODE REVIEW

File: crates/uv-interpreter/src/python_environment.rs [hunk 1]
```diff
@@ -72,6 +72,49 @@ impl PythonEnvironment {
...
+            interpreter_with_user_scheme.platlib(),
+            interpreter_with_user_scheme.purelib(),
+            interpreter_with_user_scheme.scripts(),
+            interpreter_with_user_scheme.data(),
+            interpreter_with_user_scheme.include(),
+        ];
+
+        for path in directories {
+            if !Path::new(path).exists() {
+                fs::create_dir_all(path)?;
```
Comment (konstin) [inline crates/uv-interpreter/src/python_environment.rs:108]:
```suggestion
                fs_err::create_dir_all(path)?;
```
We use `fs_err` over `std::fs` so errors have paths attach

**Comment by charliermarsh** [comment]
This looks like a nice PR! The main blocker is that I need to decide if we want to support `--user`. I think it leads to some limitations in `pip` but I just haven't had time to investigate it. For example, it was suggested elsewhere that `--user` and `--editable` are no longer supported in `pip` -- is that true?

File: crates/uv-interpreter/src/interpreter.rs [hunk 2]
```diff
@@ -498,6 +447,106 @@ impl InterpreterInfo {
 
         Ok(info)
     }
+
+    /// Return the [`InterpreterInfo`] for the given Python interpreter with user scheme.
+    pub(crate) fn query_user_scheme_info(interpreter: &Path) -> Result<Self, Error> {
+        let script = include_str!("get_interpreter_info.py");
+        let envs = vec![("_UV_USE_USER_SCHEME", "1")];
+        let output = Self::execute_script(interpreter, script, Some(envs))?;
```
Comment (charliermarsh) [inline crates/uv-interpreter/src/interpreter.rs:455]: Does this intentionally _not_ cache?

**Comment by charliermarsh** [comment]
Again, just want to say: this is very nice work! In fact I'm more inclined to support this now that we have a nice PR for it. But I do want to understand the implications a little more first.

File: crates/uv/src/commands/pip_install.rs [hunk 3]
```diff
@@ -138,6 +142,18 @@ pub(crate) async fn pip_install(
         }
     }
 
+    // Check if virtualenv has access to the system site-packages
+    // <https://github.com/pypa/pip/blob/a33caa26f2bf525eafb4ec004f9c1bd18d238a31/src/pip/_internal/commands/install.py#L686>
+    if user
+        && venv.interpreter().is_virtualenv()
+        && !venv.cfg().unwrap().include_system_site_packages()
```
Comment (charliermarsh) [inline crates/uv/src/commands/pip_install.rs:149]: Can we use `is_some_and` or `unwrap_or` here instead of unwrapping?

**Comment by imfing** [comment]
Thank you for the comments 😄 

> For example, it was suggested elsewhere that `--user` and `--editable` are no longer supported in `pip` -- is that true?

It seems there might be some confusion stemming from an issue discussed here: https://github.com/astral-sh/uv/issues/2077#issuecomment-1971579880 

> One of the problems is that you cannot run --user and --editable any more with pip. 

As far as i know, both `--user` and `--editable` are still supported in the latest [pip install options](https://pip.pypa.io/en/stable/cli/pip_install/#options). 
There appears to be no intention from the pip developers to deprecate these flags. I couldn't find any discussion about this either.

I don't have the full context so the complexity might arise specifically in the context of using Airflow. It would be great if the author of that comment could provide some further details.

For reference, I conducted a quick local test combining the `--user` and `--editable` flags:
```
❯ python3 -m venv venv --system-site-packages
❯ source venv/bin/activate
❯ pip install --user -e sampleproject/
Obtaining
  Installing build dependencies ... 
...
Successfully built sampleproject
Installing collected packages: peppercorn, sampleproject
...
Successfully installed peppercorn-0.6 sampleproject-3.0.0

❯ pip list --user
Package       Version Editable project location
------------- ------- -----------------------------------------
peppercorn    0.6
sampleproject 3.0.0   /Users/me/sampleproject
```

**Comment by zanieb** [comment]
There's some more discussion on this active in https://github.com/astral-sh/uv/issues/2077

**Comment by charliermarsh** [comment]
I may expand the section on `--user` in `PIP_COMPATIBILITY.md` in the interim.

**Comment by Malix-off** [comment]
Is there any blocker that requires some help to merge this PR?
We are waiting for that feature to be merged to migrate every devcontainer we have to use `uv`

**Comment by imfing** [comment]
> Is there any blocker that requires some help to merge this PR? We are waiting for that feature to be merged to migrate every devcontainer we have to use `uv`

In #2077, there were extensive discussions, and from what I gathered, it seemed the `uv` team prefer advocating for the adoption of virtual environments in `uv`.

I kept this PR open in case things change, and if the `uv` team is okay with adding the flag, I will update the PR.

**Comment by Malix-off** [comment]
Well then, [installing in immutable containers (such as codespaces) would require an useless venv, as `--system` doesn't work](https://github.com/astral-sh/uv/issues/2077#issuecomment-2098624319)

**Comment by zanieb** [comment]
@Malix-off can you please discuss this on the issue instead of this pull request? I'd prefer to keep the discussion here focused on implementation details. Thanks.

**Comment by Malix-off** [comment]
@zanieb absolutely, good practice btw and sorry about those annoyances

**Comment by imfing** [comment]
Noticed that `--target` flag was added in https://github.com/astral-sh/uv/pull/3257, which can be used as a workaround for `--user` by providing the user-specific [site-packages directory](https://docs.python.org/3/library/site.html#site.getusersitepackages)

**Pull request closed without merging**